esp/clients/wsdfuaccess/wsdfuaccess.cpp

```diff
@@ -59,21 +89,667 @@ bool getFileAccess(StringBuffer &metaInfo, const char *serviceUrl, const char *j
+        if (0 == binLayout.length())
+            throw makeStringExceptionV(0, "lookupDFUFile(%s) - layout missing", logicalName);
+        ret->queryEngineInterface()->setLayoutBin(binLayout.length(), binLayout.bytes());
+    }
+    return ret.getClear();
+}
+
+static IClientDFUFileCreateResponse *doCreateDFUFileDeprecated(IClientWsDfu *dfuClient, const char *logicalName, const char *cluster, DFUFileType type, const char *recDef, const char *requestId, unsigned expirySecs)
+{
+    Owned<IClientDFUFileCreateRequest> dfuReq = dfuClient->createDFUFileCreateRequest();
+
+    dfuReq->setECLRecordDefinition(recDef);
+    IEspDFUFileAccessRequestBase &requestBase = dfuReq->updateRequestBase();
+    requestBase.setName(logicalName);
+    requestBase.setCluster(cluster);
+    requestBase.setExpirySeconds(expirySecs);
+    requestBase.setJobId(requestId);
+    requestBase.setAccessRole(CFileAccessRole_External);
+    requestBase.setAccessType(CSecAccessType_Write);
+    requestBase.setReturnBinTypeInfo(true);
 
-    metaInfo.append(dfuResp->getAccessInfo().getMetaInfoBlob());
-    return true;
+    return dfuClient->DFUFileCreate(dfuReq);
 }
 
+static IClientDFUFileCreateResponse *doCreateDFUFile(IClientWsDfu *dfuClient, const char *logicalName, const char *cluster, DFUFileType type, const char *recDef, const char *requestId, unsigned expirySecs)
+{
+    Owned<IClientDFUCreateFileRequest> dfuReq = dfuClient->createDFUCreateFileRequest();
+
+    dfuReq->setECLRecordDefinition(recDef);
+    dfuReq->setName(logicalName);
+    dfuReq->setCluster(cluster);
+    dfuReq->setExpirySeconds(expirySecs);
+    dfuReq->setRequestId(requestId);
+#ifdef TEST_RETURNTEXTRESPONSE
+    dfuReq->setReturnTextResponse(true);
+#endif
+
+    CDFUFileType serviceType;
+    switch (type)
+    {
+        case dft_flat:
+            serviceType = CDFUFileType_Flat;
+            break;
+        case dft_index:
+            serviceType = CDFUFileType_Index;
+            break;
+    }
+    dfuReq->setType(serviceType);
+    return dfuClient->DFUCreateFile(dfuReq);
+}
+
+static IDFUFileAccess *doCreateDFUFileHandleLegacy(const char *serviceUrl, const char *logicalName, const char *cluster, DFUFileType type, const char *recDef, const char *requestId, unsigned expirySecs, const char *user, const char *password)
+{
+    Owned<IClientWsDfu> dfuClient = createWsDfuClient();
+    dfuClient->addServiceUrl(serviceUrl);
+    dfuClient->setUsernameToken(user, password, "");
+
+    Owned<IClientDFUFileCreateResponse> dfuResp;
+    try
+    {
+        dfuResp.setown(doCreateDFUFile(dfuClient, logicalName, cluster, type, recDef, requestId, expirySecs));
+    }
+    catch (IException *e)
+    {
+        /* NB: there should really be a different IException class and a specific error code
+         * The server knows it's an unsupported method.
+         */
+        if (SOAP_SERVER_ERROR != e->errorCode())
+            throw;
+        // fall through and try deprecated method
+        e->Release();
```

Comment:
could we potentially already know the appropriate target wsdfu interface since this requires the file access ?